pr 요청_진희

Egenieee opened the pull request:
사다리 타는 로직을 더 리팩토링 해볼 수 있을 것 같은데 저 로직에서 더 이상 어떻게 줄여야 하는 지 아무리 생각해도 안떠올라서.. 일단 pr 요청합니다! 출력하는 부분도 이름이랑 실행 결과 부분이 조금 밀리는 거 같긴 한데 저것도 프론트엔드 단 문제인거 같아서 신경 안썼습니다..!

항상 수고많으세요 ㅜ 감사합니다🥲🥲

Egenieee committed:
Egenieee branch test

Egenieee committed:
구현할 기능 목록 정리

Egenieee committed:
-feat: Line class 추가

Egenieee committed:
-feat: InputView class 추가

Egenieee committed:
참가자 이름 입력 받아서 참가자 객체 생성하는 기능 추가
-feat: Controller class 추가
-feat: LadderGame class 추가
-feat: Player 객체 추가

Egenieee committed:
-feat: List<Player> -> 일급컬렉션으로 변경

Egenieee committed:
-feat: 쉼표 검증 로직 추가

Egenieee committed:
구현할 기능 목록 수정

Egenieee committed:
실행 결과로 쓰일 값 받아서 Outputs 객체 생성
-feat: Output class 추가
-feat: 일급 컬렉션 Outputs class 추가
-feat: 쉼표 검증 로직 추가
-feat: 참가자의 수와 동일한 지 검증하는 로직 추가

Egenieee committed:
-feat: 참가자 이름 5자 이하인지 검증하는 로직 추가

Egenieee committed:
docs: 구현할 기능 목록 수정

Egenieee committed:
사다리 만드는 기능 추가
-feat: 사다리의 가로 라인 객체에서 연결 부분 랜덤으로 만드는 기능 추가
-feat: 사다리 최대 높이 만큼 Line 객체 쌓는 기능 추가
-feat: 랜덤 클래스 이용해서 추가 여부 결정하는 기능 추가

Egenieee committed:
-feat: 사다리 세로 라인에 루트 없는 경우 다시 만들도록 하기 위해 세로 라인 검증 로직 추가
-feat: 사다리 출력 위해 잠시 프린트 로직 추가

Egenieee committed:
-refactor: 세로 라인 검증 로직 리팩토링

Egenieee committed:
-feat: ladder OutputView 에서 프린트하도록 기능 추가

Egenieee committed:
-feat: 사다리 위에 이름까지 프린트 하도록 기능 추가 - 문제 많음 다시 수정

Egenieee committed:
-fix: 사다리 인원 수보다 한 세로 라인 더 만들어지는 버그 수정

Egenieee committed:
-feat: 사다리 아래에 실행 결과까지 프린트 하도록 기능 추가 - 문제 많음 추후에 다시 수정

Egenieee committed:
사다리 타는 기능 추가
-feat: 사다리 타는 로직 추가
-feat: 결과 값 담을 일급컬렉션 추가

Egenieee committed:
-refactor: OutputView 사다리 출력 수정

Egenieee committed:
-refactor: Controller에서 사다리 만들던 로직 LadderGame으로 이동

Egenieee committed:
실행 결과 출력하는 기능 추가
-feat: 입력받은 player 이름으로 Player 객체 찾는 로직 추가
-feat: 반복적으로 결과볼 사람 이름 입력 받는 로직 추가
-feat: GameResult 에서 Player를 매개변수로 Output 이름 바로 받도록 하는 로직 추가
-feat: InputView, OutputView 로직 추가

Egenieee committed:
주석 제거

Egenieee committed:
결과 출력 로직 추가 & 수정
-feat: "all" 입력했을 때 전체 결과 출력하는 로직 추가
-feat: 이름이 올바른지 검증하는 로직 추가

Egenieee committed:
주석 정리

Egenieee committed:
-refactor: 메서드 이름 수정

Egenieee committed:
-refactor: 사다리 타는 로직 사다리 객체 안으로 이동

gch01410 reviewed (changes requested):
구조 엄청 깔끔한데?? 💯 💯 
다음은 테스트 코드도 작성해서 요청해줘!

Egenieee commented on src/main/java/domain/Line.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.ArrayList;
+
+public class Line {
+    private final ArrayList<Boolean> points = new ArrayList<>();
```
엇 이건 힌트에 나와있는대로 했어요..!! 이렇게 해야하는 건줄 알고..ㅎㅎㅎ

Egenieee committed:
-refactor: 인터페이스 타입으로 선언

Egenieee committed:
-refactor: name split 부분 stream 사용하여 로직 수정

Egenieee commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,130 @@
+        return false;
+    }
+
+    private boolean checkUnderFiveLetter(String names) {
+        List<String> playerNames = LadderGame.splitNames(names);
+
+        boolean isUnderFiveLetter = playerNames.stream().allMatch(name -> name.length() < PLAYER_NAME_LIMIT);
+
+        if (isUnderFiveLetter) {
+            return true;
+        }
+
+        OutputView.printInputNamesLimitErrorMessage();
+        return false;
+    }
```
이렇게 해보려고 했는데, 아무래도 Player 객체 후보단 생성 전에 검증 로직이 들어가는 게 더 나을거 같아서 이렇게 했습니다..ㅠ

Egenieee replied:
객체 생성 후에 검증해서 다시 객체를 만들게 되면, 메모리 면에서 효율적이지 않을 거라고 생각했습니다..!!

Egenieee committed:
-refactor: input 값 검증 로직 LadderGame으로 이동

Egenieee committed:
-refactor: Player, Output 객체 생성 로직 stream 으로 수정

Egenieee committed:
-refactor: Ladder 생성자 로직 stream으로 수정

Egenieee commented on src/main/java/domain/Ladder.java:
```diff
@@ -0,0 +1,71 @@
+
+    public boolean checkLadder() {
+        int rootCount;
+        int lineRowLength = lines.get(0).getLine().size();
```
어떤 중복을 제거해야 하는지 잘 모르겠습니다..ㅠ

Egenieee committed:
Ladder class 로직 수정
-refactor: 사다리 타는 로직 수정 -> 조건문 메서드로 분리
-refactor: 선언과 초기화 동시에
-refactor: 결과물 받는 로직 수정

Egenieee committed:
-refactor: Controller 수정

Egenieee committed:
-test: LineTest 작성

Egenieee committed:
-test: RandomRootCreatorTest 작성

Egenieee committed:
-refactor: RandomLadderCreator class 추가하여 사다리 만드는 로직 이동

Egenieee committed:
-test: RandomLadderCreatorTest 작성

Egenieee committed:
-refactor: Ladder class 인스턴스 변수 추가, 중복 코드 제거

Egenieee committed:
-refactor: RandomLineCreator class 추가, 랜덤 라인 만드는 로직 이동

Egenieee committed:
-test: LadderTest 작성

Egenieee committed:
-test: RandomLineCreatorTest 작성

Egenieee committed:
-refactor: Line class 중복 코드 삭제

Egenieee committed:
-test: PlayersTest 작성

Egenieee committed:
-test: OutputsTest 작성

Egenieee committed:
-test: LadderGameTest 작성

Egenieee committed:
-test: GameResultTest 작성

gch01410 commented on src/main/java/domain/LadderGame.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    public static List<String> splitNames(String names) {
+        return Arrays.stream(names.split(",")).collect(Collectors.toList());
```
보통 stream을 사용할때 한 줄에 `.` 1개만 찍어!
```java
return Arrays.stream(names.split(","))
            .collect(Collectors.toList());
```

gch01410 commented on src/main/java/domain/LadderGame.java:
```diff
@@ -0,0 +1,93 @@
+        Map<Player, Output> result = new HashMap<>();
+
+        for (int columnIndex = 0 ; columnIndex < players.getPlayerCount() ; columnIndex++) {
+            Player player = players.getPlayers().get(columnIndex);
```
Players객체에서 원하는  Player만 받아올 수 있지 않을까?

gch01410 commented on src/test/java/domain/GameResultTest.java:
```diff
@@ -0,0 +1,28 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class GameResultTest {
+    @Test
+    public void 게임_결과_생성() {
+        List<Line> lines = Arrays.asList(new Line(Arrays.asList(true, false)),
+                new Line(Arrays.asList(false, true)), new Line(Arrays.asList(true, false)));
+
```
💯 💯

gch01410 commented on src/main/java/domain/RandomLineCreator.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.List;
+
+public class RandomLineCreator {
+    private final List<Boolean> points = new ArrayList<>();
```
points를 필드로 선언하지 않아도 괜찮지 않을까?

gch01410 commented on src/test/java/domain/LadderGameTest.java:
```diff
@@ -0,0 +1,133 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class LadderGameTest {
```
<img width="882" alt="스크린샷 2022-04-28 오후 11 16 03" src="https://user-images.githubusercontent.com/26810182/165775076-1e90460b-48fb-4ab6-83ba-ca4f0b651145.png">

인텔리J 오른쪽 위에 방패모양 누르면 테스트 코드 실행시키고 테스트 커버리지 숫자가 나오거든?
특정 패키지, 클래스, 메서드, 라인의 테스트 커버리지를 알 수 있어
커버리지 👍 👍

gch01410 commented on src/test/java/domain/RandomLineCreatorTest.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.List;
+
+public class RandomLineCreatorTest {
+    @Test
+    public void 랜덤_라인_생성() {
+        int playerCount = 3;
+
+        List<Boolean> points = new RandomLineCreator().getLine(playerCount);
+        boolean isContinuousRoot = checkRoots(points);
+
+        assertThat(points.size()).isEqualTo(playerCount - 1);
+        assertThat(isContinuousRoot).isFalse();
+    }
```
https://tecoble.techcourse.co.kr/post/2020-05-17-appropriate_method_for_test_by_interface/
https://tecoble.techcourse.co.kr/post/2020-05-07-appropriate_method_for_test_by_parameter/
여기 글 읽어보고 랜덤 때문에 테스트하기 어려웠던 메서드들도 더 테스트 해보자!

gch01410 reviewed (changes requested):
잘하고 있어! 💯 
조금만 더 리팩토링해보자! 거의 다 왔다!

Egenieee committed:
-refactor: ladder height, width 구하는 로직 메서드로 분리

Egenieee commented on src/main/java/domain/Ladder.java:
```diff
@@ -0,0 +1,70 @@
+        int currentRowLine;
+        int currentColumnLine;
+
+        for (currentRowLine = startLine, currentColumnLine = 0 ; currentColumnLine < ladderHeight  ; ) {
```
맨날 헷갈리네요 이건 ㅎㅎ;;

Egenieee committed:
-refactor: 사다리 타는 로직 안의 변수명 수정

Egenieee committed:
-refactor: RandomLadderCreator 로직 수정

Egenieee committed:
-refactor: LadderGame stream 문법 코드 정리

Egenieee committed:
LadderGame의 검증 로직 이동
-refactor: Validator class 새로 추가
-test: ValidatorTest 작성

Egenieee committed:
-refactor: LadderGame 메서드 이름 수정

Egenieee committed:
Players 에서 한 사람만 리턴하도록 로직 추가
-refactor: Players 메서드 추가
-test: PlayersTest 테스트 추가

Egenieee committed:
-refactor: RandomLineCreator points 로컬 변수로 변경

Egenieee committed:
-refactor: points -> final로 변경

Egenieee committed:
-refactor: RootCreator Interface 생성

Egenieee committed:
-refactor: RandomRootCreator가 인터페이스 RootCreator를 오버라이드하도록 수정

Egenieee committed:
-refactor: RandomLineCreator 수정

Egenieee committed:
-test: RandomRootCreatorTest 수정

Egenieee committed:
-refactor: for test MovableRootCreator 생성

Egenieee committed:
-refactor: for test NonMovableRootCreator 생성

Egenieee committed:
생성한 인터페이스 사용하도록 수정
-test: RandomLineCreatorTest 수정
-test: RandomLadderCreatorTest 수정
-refactor: RandomLadderCreator 수정
-refactor: RandomLineCreator 수정
-refactor: LadderGame 수정
-test: LadderTest 수정

gch01410 commented on src/main/java/domain/Ladder.java:
```diff
@@ -0,0 +1,79 @@
+        return currentRowLine != 0 && points.get(currentRowLine - 1);
+    }
+
+    private int getLadderHeight() {
+        return lines.size();
+    }
+
+    private int getLadderWidth() {
+        return lines.get(0).getLineLength();
+    }
```
👍 👍

gch01410 commented on src/main/java/domain/Validator.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class Validator {
```
👍

gch01410 commented on src/main/java/domain/MovableRootCreator.java:
```diff
@@ -0,0 +1,9 @@
+package domain;
+
+public class MovableRootCreator implements RootCreator{
+
+    public boolean createRoot() {
+        return true;
+    }
+
+}
```
테스트 코드에서 사용할 용도로 만든 클래스는 테스트 패키지에 위치시켜도 될 것 같아!

gch01410 reviewed (approved):
3달동안 수고 많았어! 👏 👏 
머지할게~
